### PR TITLE
Add Heal command

### PR DESCRIPTION
**Pull request by pietrodimarco-dfinity**
This introduces the `heal` command, designed to improve the health of the network by replacing unhealthy nodes within specified subnets. 
The command traverses all nodes in the network, identifying and replacing the unhealthy in each subnet. 

Users can optionally specify the argument `--max-replaceable-nodes-per-sub` to optimize the healing process up to max-replaceable-nodes-per-sub nodes. 
If the `--max-replaceable-nodes-per-sub` argument is not provided, the algorithm will proceed replacing all and only the unhealthy nodes in the subnets.

### CODE REVIEW

File: rs/cli/src/runner.rs [hunk 1]
```diff
@@ -444,4 +445,30 @@ impl Runner {
...
+    pub async fn network_heal(
+        &self,
+        request: ic_management_types::requests::HealRequest,
+        _verbose: bool,
+        simulate: bool,
+    ) -> Result<(), anyhow::Error> {
+        let change = self.dashboard_backend_client.network_heal(request).await?;
+        println!("{}", change);
+
+        join_all(change.subnets_change_response.iter().map(|subnet_change_response| async move {
```
Comment (sasa-tomic) [inline rs/cli/src/runner.rs:458]: So we're collecting the replacement results in parallel? Is this time consuming, or why do we do it in parallel? I don't mind, I'm just curious.

File: rs/decentralization/src/network.rs [hunk 2]
```diff
@@ -985,3 +985,78 @@ impl Display for SubnetChange {
...
+
+impl NetworkHealRequest {
+    pub fn new(subnets: Vec<NetworkHealSubnets>) -> Self {
+        Self { subnets }
+    }
+
+    pub fn heal_and_optimize(
+        &self,
+        mut available_nodes: Vec<Node>,
+        max_replacable_nodes: Option<usize>,
```
Comment (sasa-tomic) [inline rs/decentralization/src/network.rs:1031]:
```suggestion
        max_replaceable_nodes: Option<usize>,
```

File: rs/cli/src/cli.rs [hunk 3]
```diff
@@ -57,6 +57,15 @@ pub enum Commands {
         /// Path to the DER file
         path: String,
     },
+
+    Heal {
+        /// Max number of nodes to be replaced per subnet.
+        /// Optimization will be performed automatically maximizing the decentralization
+        /// and minimizing the number of replaced nodes per subnet
+        #[clap(short, long)]
+        max_replacable_nodes_per_sub: Option<usize>,
```
Comment (sasa-tomic) [inline rs/cli/src/cli.rs:66]:
```suggestion
        max_replaceable_nodes_per_sub: Option<usize>,
```

File: rs/decentralization/src/network.rs [hunk 4]
```diff
@@ -985,3 +985,78 @@ impl Display for SubnetChange {
...
+    ) -> Result<Vec<SubnetChangeResponse>, NetworkError> {
+        let mut subnets_changed = Vec::new();
+        let subnets_to_heal = self.subnets.iter().sorted_by(|a, b| a.cmp(b).reverse()).collect_vec();
+
+        for subnet in subnets_to_heal {
+            let unhealthy_nodes_len = subnet.unhealthy_nodes.len();
+            if let Some(max_replacable_nodes) = max_replacable_nodes {
```
Comment (sasa-tomic) [inline rs/decentralization/src/network.rs:1038]:
```suggestion
            if let Some(max_replaceable_nodes) = max_replaceable_nodes {
```

File: rs/decentralization/src/network.rs [hunk 5]
```diff
@@ -985,3 +985,78 @@ impl Display for SubnetChange {
...
+                if unhealthy_nodes_len > max_replacable_nodes {
+                    warn!(
+                        "Subnet {} has {} unhealthy nodes\nMax replacable nodes is {} skipping...",
```
Comment (sasa-tomic) [inline rs/decentralization/src/network.rs:1041]:
```suggestion
                        "Subnet {} has {} unhealthy nodes\nMax replaceable nodes is {} skipping...",
```

File: rs/decentralization/src/network.rs [hunk 6]
```diff
@@ -985,3 +985,78 @@ impl Display for SubnetChange {
...
+        mut available_nodes: Vec<Node>,
+        max_replacable_nodes: Option<usize>,
+    ) -> Result<Vec<SubnetChangeResponse>, NetworkError> {
+        let mut subnets_changed = Vec::new();
+        let subnets_to_heal = self.subnets.iter().sorted_by(|a, b| a.cmp(b).reverse()).collect_vec();
+
+        for subnet in subnets_to_heal {
+            let unhealthy_nodes_len = subnet.unhealthy_nodes.len();
+            if let Some(max_replacable_nodes) = max_replacable_nodes {
+                if unhealthy_nodes_len > max_replacable_nodes {
```
Comment (sasa-tomic) [inline rs/decentralization/src/network.rs:1039]:
```suggestion
                if unhealthy_nodes_len > max_replaceable_nodes {
```

File: rs/ic-management-types/src/requests.rs [hunk 7]
```diff
@@ -114,3 +114,8 @@ impl NodeRemovalReason {
         }
     }
 }
+
+#[derive(Serialize, Deserialize)]
+pub struct HealRequest {
+    pub max_replacable_nodes_per_sub: Option<usize>,
```
Comment (sasa-tomic) [inline rs/ic-management-types/src/requests.rs:120]:
```suggestion
    pub max_replaceable_nodes_per_sub: Option<usize>,
```

File: rs/ic-management-backend/src/endpoints/network.rs [hunk 8]
```diff
@@ -0,0 +1,36 @@
...
+            Ok::<NetworkHealSubnets, NetworkError>(NetworkHealSubnets {
+                name: unhealthy_subnet.metadata.name.clone(),
+                decentralized_subnet: DecentralizedSubnet::from(unhealthy_subnet),
+                unhealthy_nodes,
+            })
+        })
+        .collect_vec();
+
+    let subnets_change_response =
+        NetworkHealRequest::new(subnets_to_heal).heal_and_optimize(registry.available_nodes().await?, request.max_replacable_nodes_per_sub)?;
```
Comment (sasa-tomic) [inline rs/ic-management-backend/src/endpoints/network.rs:33]:
```suggestion
        NetworkHealRequest::new(subnets_to_heal).heal_and_optimize(registry.available_nodes().await?, request.max_replaceable_nodes_per_sub)?;
```

File: rs/decentralization/src/network.rs [hunk 9]
```diff
@@ -985,3 +985,78 @@ impl Display for SubnetChange {
...
+                    warn!(
+                        "Subnet {} has {} unhealthy nodes\nMax replacable nodes is {} skipping...",
+                        subnet.decentralized_subnet.id, unhealthy_nodes_len, max_replacable_nodes
+                    );
+                    continue;
+                }
+            }
+            let optimize_limit = max_replacable_nodes.unwrap_or(unhealthy_nodes_len) - unhealthy_nodes_len;
```
Comment (sasa-tomic) [inline rs/decentralization/src/network.rs:1047]:
```suggestion
            let optimize_limit = max_replaceable_nodes.unwrap_or(unhealthy_nodes_len) - unhealthy_nodes_len;
```

File: rs/decentralization/src/network.rs [hunk 10]
```diff
@@ -985,3 +985,78 @@ impl Display for SubnetChange {
...
+        let mut subnets_changed = Vec::new();
+        let subnets_to_heal = self.subnets.iter().sorted_by(|a, b| a.cmp(b).reverse()).collect_vec();
+
+        for subnet in subnets_to_heal {
+            let unhealthy_nodes_len = subnet.unhealthy_nodes.len();
+            if let Some(max_replacable_nodes) = max_replacable_nodes {
+                if unhealthy_nodes_len > max_replacable_nodes {
+                    warn!(
+                        "Subnet {} has {} unhealthy nodes\nMax replacable nodes is {} skipping...",
+                        subnet.decentralized_subnet.id, unhealthy_nodes_len, max_replacable_nodes
```
Comment (sasa-tomic) [inline rs/decentralization/src/network.rs:1042]:
```suggestion
                        subnet.decentralized_subnet.id, unhealthy_nodes_len, max_replaceable_nodes
```

File: rs/decentralization/src/network.rs (hunk 6 above)
Comment (sasa-tomic) [reply rs/decentralization/src/network.rs:1039]: Would it be possible to replace the `max_replaceable_nodes` even though it's less than the number of unhealthy nodes? Maybe picking first N (==max_replaceable_nodes) nodes and replacing them?

**Review by sasa-tomic — COMMENTED** [review]
Looks great! Thank you!

File: rs/cli/src/runner.rs (hunk 1 above)
Comment (pietrodimarco-dfinity) [reply rs/cli/src/runner.rs:458]: Well actually the only reason why I am doing the join_all is to execute the run_membership_change inside the map...it is actually sequential execution of the proposals.

File: rs/decentralization/src/network.rs [hunk 11]
```diff
@@ -985,3 +985,85 @@ impl Display for SubnetChange {
...
+            if let Some(max_replaceable_nodes) = max_replaceable_nodes {
+                if unhealthy_nodes_len > max_replaceable_nodes {
+                    unhealthy_nodes = subnet.unhealthy_nodes.clone().into_iter().take(max_replaceable_nodes).collect_vec();
+
+                    warn!(
+                        "Subnet {} has {} unhealthy nodes\nMax replacable nodes is {}\nNodes considered for replacement: {:?}",
+                        subnet.decentralized_subnet.id,
+                        unhealthy_nodes_len,
+                        max_replaceable_nodes,
+                        unhealthy_nodes.iter().map(|node| node.id).collect_vec()
```
Comment (sasa-tomic) [inline rs/decentralization/src/network.rs:1049]:
```suggestion
                        "Subnet {}: replacing {} of {} unhealthy nodes: {:?}",
                        subnet.decentralized_subnet.id,
                        max_replaceable_nodes,
                        unhealthy_nodes_len,
                        unhealthy_nodes.iter().map(|node| node.id).collect_vec()
```

**Review by sasa-tomic — APPROVED** [review]
Awesome!

File: rs/cli/src/runner.rs [hunk 12]
```diff
@@ -444,4 +445,30 @@ impl Runner {
             .await?;
         Ok(())
     }
+
+    pub async fn network_heal(
+        &self,
+        request: ic_management_types::requests::HealRequest,
+        _verbose: bool,
```
Comment (NikolaMilosa) [inline rs/cli/src/runner.rs:452]: Unused argument

File: rs/cli/src/runner.rs [hunk 13]
```diff
@@ -444,4 +445,30 @@ impl Runner {
...
+                ops_subnet_node_replace::replace_proposal_options(subnet_change_response)?,
+                simulate,
+            )
+            .await
+            .map_err(|e| {
+                println!("{}", e);
+                e
+            })
+        }))
+        .await;
```
Comment (NikolaMilosa) [inline rs/cli/src/runner.rs:470]:
```suggestion
       let errors = join_all(change.subnets_change_response.iter().map(|subnet_change_response| async move {
            self.run_membership_change(
                subnet_change_response.clone(),
                ops_subnet_node_replace::replace_proposal_options(subnet_change_response)?,
                simulate,
            )
            .await
            .map_err(|e| {
                println!("{}", e);
                e
            })
        }))
        .await
        .into_iter()
        .filter_map(|f| f.err())
        .collect::<Vec<_>>();

        if !errors.is_empty() {
            anyhow::bail!("Errors: {:?}", errors);
        }
```

File: rs/decentralization/src/network.rs [hunk 14]
```diff
@@ -985,3 +985,85 @@ impl Display for SubnetChange {
...
+                    );
+                }
+            }
+            let unhealthy_nodes_len = unhealthy_nodes.len();
```
Comment (NikolaMilosa) [inline rs/decentralization/src/network.rs:1053]: Seems like a copy of line 1038

File: rs/ic-management-backend/src/subnets.rs [hunk 15]
```diff
@@ -2,7 +2,44 @@ use std::collections::BTreeMap;
...
+                            None
+                        } else {
+                            info!("Node {} is {:?}", n.principal, health);
+                            Some(n)
+                        }
+                    }
+                    None => {
+                        warn!("Node {} has no known health, assuming unhealthy", n.principal);
+                        Some(n)
+                    }
```
Comment (NikolaMilosa) [inline rs/ic-management-backend/src/subnets.rs:31]: AFAIK we run backend as a child process and we can't really see the logs coming from it so the `info!` and `warn!` aren't really visible right? If they are not visible I would just write it simpler:
```suggestion
                    Some(health) if *health == ic_management_types::Status::Healthy => None,
                    _ => Some(n) 
```

**Review by NikolaMilosa — APPROVED** [review]
Noice job machine 🤖

File: rs/cli/src/runner.rs (hunk 13 above)
Comment (NikolaMilosa) [reply rs/cli/src/runner.rs:470]: For consistency, I would skip a `println!` here and just bail on errors and print them as a return of the whole command

File: rs/decentralization/src/network.rs (hunk 14 above)
Comment (pietrodimarco-dfinity) [reply rs/decentralization/src/network.rs:1053]: This is necessary to update again the value

**Pull request merged**